5주차 과제 - 레스토랑 조회 구현하기

sejineeee opened the pull request:
## 내용

안녕하세요!!
리덕스와 테스트가 아직 익숙하지 않은 상태에서 비동기를 하려고 하니 어렵네요! 😯
작업하는대로 바로바로 커밋해서 올리도록 하겠습니다!!

sejineeee committed:
feat: redux store 파일 생성

createStore 이용하여 store 생성
redux thunk에 middleWare 설정

sejineeee committed:
Provider 설정

sejineeee committed:
feat: api 파일 추가

categories와 regions의 데이터를 받아올 수 있는 api 호출 함수 구현

sejineeee committed:
Regions 테스트용 데이터 생성

fixtures 폴더를 만들어서 regions 테스트용 데이터 생성

sejineeee committed:
test: redux 테스트를 위해서 mock 생성

useDispatch, useSelector mock 파일 생성

sejineeee committed:
App 파일에 RegionsContainer 컴포넌트 추가

sejineeee committed:
test: RegionsContainer 테스트 구현

RegionsContainer가 렌더링 됐을 때 text 서울을 갖고 있는지 테스트

sejineeee committed:
feat: action 함수 구현

setRegions, loadRegions action 구현

sejineeee committed:
feat: reducer 구현

reducers 객체에 setRegions 메소드 구현
initialState 설정해서 reducer 함수에 state 기본값으로 할당
action이 없거나 action type이 일치하는게 없을 경우 state 반환
그 외에는 reducers 객체에 action type에 일치하는 메소드 호출

sejineeee committed:
test: reducer 테스트 구현

setRegions action에 대해서 테스트

sejineeee committed:
feat: RegionsContainer 구현

useEffect를 이용하여 마운트 됐을 때, dispatch에 loadRegions 함수 전달
Regions 컴포넌트에 props regions 전달

sejineeee committed:
test: Regions test 구현

Regions 컴포넌트 test 구현
컴포넌트가 렌더링 되었을 때, 서울과 대전 텍스트를 갖고 있는지 테스트

sejineeee committed:
feat: Regions 컴포넌트 구현

props로 전달받은 데이터를 map을 이용하여 구현

sejineeee committed:
test: App 테스트 구현

RegionsContainer가 구현된 상태에서의 App 테스트
App이 렌더링되었을 때, 해당 텍스트를 갖고 있는지 테스트

sejineeee committed:
테스트용 categories 파일 생성

테스트를 하기 위한 데이터 categories 파일 생성

daadaadaah commented on src/reducer.test.js:
```diff
@@ -0,0 +1,18 @@
+
+describe('reducer', () => {
+  describe('setRegions', () => {
+    it('지역에 대한 데이터를 받아온다.', () => {
```
reducer 관점에서 표현해주시는게 좋아요! 테스트 대상이 Reducer 이니까요!

daadaadaah reviewed (changes requested):
한주한주 새로운 개념을 익히시는거 어렵죠ㅠㅠ이전 개념도 여전히 새로운대, 거기에 추가로 또 새로운 개념을 익혀야하니ㅜㅜ그래도, 하나하나 차근차근 하다보면, 금새 익숙해지실 겁니당!ㅎㅎ 다음 PR 기다리고 있을게요~🚀

sejineeee committed:
test: CategoriesContainer 테스트 구현

CategoriesContainer 렌더링 됐을 때 테스트 구현

sejineeee committed:
categories에 대한 action 추가

setCategories action 구현
loadCategories 함수 구현
fetchCategories import 하여 loadCategories 함수에서 async await 사용하여 호출

sejineeee committed:
feat: CategoriesContainer 구현

CategoriesContainer 구현
useEffect 사용하여 마운트 됐을 때, dispatch에 loadCategories 전달
Categories 컴포넌트 호출하여 props로 전달

sejineeee committed:
setCategories 추가

reducers에 setCategories 추가

sejineeee committed:
test: Categories 컴포넌트 테스트 구현

Categories 컴포넌트가 렌더링 되었을 때 테스트 구현

sejineeee committed:
feat: Categories 컴포넌트 구현

props로 전달받은 categories 데이터를 map을 이용하여 구현
현재 클릭한 것에 대해서 확인하기 위해서 handleClick 함수에서 콘솔 출력

sejineeee committed:
test: App 테스트에 Categories 추가

categories fixtures 테스트용 데이터 중 한식이 있는지 테스트

sejineeee committed:
App 파일에 CategoriesContainer 컴포넌트 추가

sejineeee committed:
Container 컴포넌트에 handleClick 함수 추가

각 Container 컴포넌트에 handleClick 함수 추가
handleClick 함수 props 전달
handleClick 함수에서 각 id, name을 확인차 콘솔 출력

sejineeee committed:
각 컴포넌트에 onClick 이벤트 추가

props로 전달받은 onClick을 onClick 이벤트에 활용
id, name을 인자로 전달

sejineeee commented:
## 내용
현재 Regions, Categories api 호출 함수를 구현해놓은 상태입니다!
그리고 지금은 해당 버튼을 눌렀을 때, 선택된 정보를 store에 저장하기 위한 액션을 구현하는 중입니다.
아직 테스트는 덜 구현된 상태입니다...😭

daadaadaah commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,17 @@
+
+    expect(getByText('한식')).not.toBeNull();
+  })
+})
```
한줄 띄어주세요!

daadaadaah reviewed (changes requested):
좋습니다! 나머지 구현하시고 또 PR 날려주세요👊

sejineeee committed:
test: Categories 테스트에 클릭했을 때와 체크표시 테스트 추가

카테고리를 클릭했을 때, id 값을 전달하는 것을 테스트
카테고리를 클릭했을 때, 선택된 카테고리에 체크 표시가 붙는 것 테스트
renderCategories에 인자로 id 값을 전달하여 확인하고 싶었으나,
fireEvent가 실행되기 이전에 함수가 호출되는 사항으로 테스트 오작동

sejineeee committed:
선택된 카테고리 체크표시 구현

단락평가를 이용해서 selectedCategory와 category.id가 같을 때 체크표시 반환

sejineeee committed:
test: 카테고리 클릭했을 때 dispatch 실행 테스트 추가

카테고리 클릭했을 때, dispatch가 호출되는지 테스트 구현

sejineeee committed:
handleClick 함수 구현, store에 저장된 selectedCategory props로 전달

handleClick 함수에 dispatch 호출, dispatch에 selectedCategory 액션 전달
selectedCategory action에 선택된 id값을 인자로 전달
store에 저장된 선택된 id 값을 Categories 컴포넌트에 selectedCategory props로 전달

sejineeee committed:
test: 지역을 클릭하면 dispatch 호출 테스트 추가

지역을 클릭했을 때, dispatch가 호출되는지 테스트 추가

sejineeee committed:
handleClick 함수 구현, 선택된 지역 Regions 컴포넌트에 props로 전달

handleClick에 dispatch 호출하여 selectedRegion action에 name을 인자로 전달
store에 저장된 selectedRegion을 selectedRegionName으로 꺼내와 Regions에 props로 전달

sejineeee committed:
test: 지역을 선택했을 때, 선택된 지역에 체크 표시 테스트 추가

선택된 지역에 체크 표시 테스트 구현하였으나 실패
getByText에 대전을 찾을 수 없음.
클릭 이벤트 전에 체크 표시가 붙은 값 반환으로 인해 테스트 오동작

sejineeee committed:
지역 선택시 체크표시 되도록 구현

selectedRegionName을 Props로 전달 받음
단락평가 이용하여 selectedName과 region.name이 같으면 체크 표시 반환

sejineeee committed:
regionName과 categoryId를 매개변수로 받아서 각 쿼리스트링으로 전달

regionName, categoryId를 매개변수로 받아서 api url에 쿼리스트링으로 전달
선택된 지역, 카테고리에 맞는 api 호출 함수 구현

sejineeee committed:
initialState에 selected key 추가, selected 관련한 action creator 추가

initialState에 선택된 레스토랑 목록을 받을 수 있는 selectedRestaurant 추가
initialState에 선택된 지역명을 받을 수 있는 selectedRegion 추가
initialState에 선택된 카테고리를 받을 수 있는 selectecCategory를 null 값 할당하여 추가
selectedRegion, selectedCategory, selectedRestaurant 각 action creator를 reducers에 추가

sejineeee committed:
test: review 사항 반영하여 테스트명 변경, setCategories 액션에 대한 테스트 추가

review 사항 반영하여 리듀서 관점에서 테스트명 받아온다에서 저장한다로 변경
setCategories 액션에 대한 테스트 추가
selectedRegion 액션은 현재 진행 중인 상태이므로 완성한 후 다시 푸시할 예정

sejineeee committed:
App에서 RestaurantContainer 컴포넌트 호출

sejineeee committed:
선택된 name, category에 대한 action 추가, loadRestaurant 함수 구현

selectedRegion, selectedCategory, selectedRestaurant action 추가
loadRestaurant 함수에서 async await 이용하여 fetchSelected 함수 호출하여 데이터 받아옴
dispatch 호출하여 selectedRestaurant 액션에 받아온 데이터 전달

sejineeee committed:
feat: RestaurantContainer 구현

RestaurantContainer 컴포넌트 생성
useEffect 이용, 의존성에 selectedRegion과 selectedCategory 추가
selectedRegion과 selectedCategory 둘 중 하나라도 초기값과 같을 경우 평가 종료
selectedRegion, selectedCategory가 둘 다 있을 경우 dispatch 호출
Restaurant 컴포넌트에 props로 restaurant 리스트 전달

sejineeee committed:
feat: Restaurant 컴포넌트 생성

전달 받은 restaurants의 length가 0일 경우 레스토랑 정보가 존재하지 않다 반환
restaurants에 데이터가 있을 경우 map 활용

sejineeee committed:
test: 테스트용 데이터 restaurants 생성

Restaurants를 테스트하기 위한 데이터 restarants 배열 생성

sejineeee commented:
## 내용
아직 RestaurantContainer, Restaurant, api에 대한 테스트를 진행하지 못한 상황입니다!
reducer에 대한 테스트도 아직 완성되지 못한 상황입니다!
test 파일을 먼저 작성하고, 컴포넌트를 작성해야 되지만... 계속 하다 보면, 저도 모르게 컴포넌트 파일을 먼저 만들고 있습니다 😭
그러고 컴포넌트 파일에 맞게 테스트를 맞추게 돼요ㅠㅠ

## 질문
<img width="1004" alt="스크린샷 2022-11-11 오후 5 57 00" src="https://user-images.githubusercontent.com/86041335/201305524-508b16cd-7e4b-4aee-a23b-0e9211b7230d.png">
<img width="585" alt="스크린샷 2022-11-11 오후 5 58 24" src="https://user-images.githubusercontent.com/86041335/201305534-88691c29-3739-4822-8e96-577bc6773967.png">

"지역과 카테고리를 선택했을 때,  선택된 값에 체크표시가 붙는다" 이렇게 테스트 한 부분이 제 의도와 다르게 테스트가 되어서 고민입니다.
하나씩 다른 방식으로 테스트 해보았는데, 둘 다 동일하게 오작동을 해서 어떤 부분을 고쳐야 될 지 잘 모르겠습니다!
`fireEvent.click` 전에 `not.toBeCalled()`를 해줬는데, 계속 함수가 호출되어서 흠,,, 이게 테스트 코드를 잘못 구현한 것인지 컴포넌트를 잘못 구현한 것인지 잘 모르겠습니다ㅠㅠㅠㅠ

---
추가적으로 토요일날 작업하여 올렸습니다. 하지만 RestaurantContainer에서 선택된 지역과 카테고리가 있을 경우 dispatch가 호출되는지 테스트 해주고 싶었으나 
![스크린샷 2022-11-12 오후 4 54 43](https://user-images.githubusercontent.com/86041335/201464623-70b67b05-0242-46df-8c68-21b1e68ac3a2.png)
이렇게 오류가 나타난 상태입니다! useSelector 부분을 잘못한 것인지 잘 모르겠습니다...😭

sejineeee committed:
test: Restaurant 테스트 구현

레스토랑 정보가 있을 경우와 없을 경우를 context로 구분하여 테스트 구현

sejineeee committed:
테스트용 데이터 간략하게 수정

필요없는 정보는 지우고, 간략하게 데이터 수정

sejineeee committed:
test: RestaurantContainer 테스트

현재 테스트 작업 진행중
선택된 지역이나 카테고리가 있을 경우와 없을 경우 context로 구분
dispatch 작동 여부에 대한 테스트를 진행할 예정

sejineeee committed:
selectedCategory 기본값 null에서 ''으로 변경

sejineeee committed:
useEffect내 조건 변경 및 오타 수정

selectedCategory 기본값 변경으로 useEffect내 조건 변경
import 할 때 Restaurant 오타 수정

sejineeee committed:
test: Restaurant 테스트 구현

선택된 지역과 카테고리가 있을 경우 dispatch 호출에 대해 테스트
하지만, 선택된 지역과 카테고리가 있을 경우 테스트 오류

daadaadaah commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,38 @@
+    expect(getByText('대전')).not.toBeNull();
+  });
+
+  it('지역을 클릭하면 선택된 지역에 체크 표시가 붙는다.', () => {
+    const { getByText } = renderRegions({
+      regionLists: regions,
+      selectedRegionName: '대전',
+    });
+
+    fireEvent.click(getByText('대전'));
+
+    expect(handleClick).toBeCalled();
+    expect(getByText('대전✅')).not.toBeNull();
+  });
```
Regions의 관심사가 무엇일까요?
클릭에 따른 상태 변화를 담당하는 컴포넌트 인가요? 아니면 props로 내려준 상태를 보여주는 컴포넌트인가요?

sejineeee replied:
props로 내려준 상태를 보여주는 컴포넌트입니다!!

daadaadaah replied:
그럼, props로 무엇을 내려주었고, 그로 인해 UI가 어떻게 보이는지에 대해 테스트하셔야 할 것 같아요!
지금 selectedRegionName으로 '대전'을 내려주면, '대전✅' 로 보여서 '대전'을 클릭할 수 없지 않을까요?

daadaadaah commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,49 @@
+    expect(container).toHaveTextContent('레스토랑 정보가 존재하지 않습니다.');
+  });
+
+  context('선택된 지역이나 카테고리가 없을 경우', () => {
+    useSelector.mockImplementation((selector) => selector({
+      selectedRegion: '',
+      selectedCategory: '',
+      selectedRestaurant: [],
+    }));
+
+    it('아무것도 하지 않는다.', () => {
+      expect(dispatch).not.toBeCalled();
+    });
+  });
```
it문에 useSelector와 renderRestaurantContainer() 넣어주시고 한번 테스트해보세요!(밑에 테스트도 같이 수정해주세요!)

daadaadaah commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,49 @@
+    return render((<RestaurantContainer />));
+  }
+
+  it('RestaurantContainer가 렌더링된다.', () => {
+    const { container } = renderRestaurantContainer();
+
+    expect(container).toHaveTextContent('레스토랑 정보가 존재하지 않습니다.');
```
여기에는 useSelector가 빠져있네용? 어느 상황에서 테스트 하고 싶으셨나요?

daadaadaah commented on src/reducer.test.js:
```diff
@@ -0,0 +1,38 @@
+
+describe('reducer', () => {
+  describe('setRegions', () => {
+    it('지역 데이터를 저장한다.', () => {
```
describe와 It문에 나온 내용을 한 문장으로 만들었을 때 자연스러운 문장이 되도록 작성해주시는게 좋아요! 지금 상황에서 `setRegions는 지역 데이터를 저장한다.` 로 읽히는데, `setRegions는 지역 목록을 변경한다.`가 좀더 적절하지 않을까 하는데, 어떻게 생각하시나여?

daadaadaah commented on src/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,30 @@
+export default function RestaurantContainer() {
+  const dispatch = useDispatch();
+
+  const { selectedRegion, selectedCategory } = useSelector((state) => ({
+    selectedRegion: state.selectedRegion,
+    selectedCategory: state.selectedCategory,
+  }));
```
24번줄이랑 통합해서 하나의 useSelector로 가져오셔도 됩니다!ㅎㅎ

daadaadaah commented on src/RegionsContainer.jsx:
```diff
@@ -0,0 +1,32 @@
+
+  const { regions, selectedRegionName } = useSelector((state) => ({
+    regions: state.regions,
+    selectedRegionName: state.selectedRegion,
```
용어를 통일해주시는게 좋아요!

daadaadaah reviewed (changes requested):
> test 파일을 먼저 작성하고, 컴포넌트를 작성해야 되지만... 계속 하다 보면, 저도 모르게 컴포넌트 파일을 먼저 만들고 있습니다 😭

- 아샬님도 테스트 파일이랑 구현 파일 왔다갔다 하면서 TDD 하십니다! 다만, 지금은 훈련중이므로, 의도적으로 테스트 먼저 작성해보시는 걸 권해드립니다!ㅎㅎ이게 습관이 되면, 언제 테스트 먼저 작성해야하는지, 언제 구현 코드 먼저 작성해야하는지 선택적으로 할 수 있기 때문이예요!ㅎㅎ

> "지역과 카테고리를 선택했을 때, 선택된 값에 체크표시가 붙는다" 이렇게 테스트 한 부분이 제 의도와 다르게 테스트가 되어서 고민입니다.
- 리뷰에 관련 내용 남겨드렸습니다!
> 추가적으로 토요일날 작업하여 올렸습니다. 하지만 RestaurantContainer에서 선택된 지역과 카테고리가 있을 경우 dispatch가 호출되는지 테스트 해주고 싶었으나 이렇게 오류가 나타난 상태입니다! useSelector 부분을 잘못한 것인지 잘 모르겠습니다...😭
- 리뷰에 관련 내용 남겨드렸습니다!

sejineeee committed:
test: api 테스트를 위해서 api mock 생성

sejineeee committed:
it 구문에 useSelector, render 추가, clearAllMocks 추가

sejineeee committed:
review: 이름 통일시키기 위해서 selectedRegionName으로 변경

sejineeee committed:
review: 이름을 통일시키기 위해서 selectedRegionName으로 변경

sejineeee committed:
review: 이름을 통일시키기 위해 변경

selectedRegionName, selectedCategoryId 이름 변경

sejineeee committed:
review: 클릭했을 때가 아닌 UI가 그려지는 것을 테스트

클릭했을 때 체크 표시가 아닌, UI가 그려지는 것으로 테스트 수정

sejineeee committed:
테스트에 restaurants 빈 배열 추가

sejineeee committed:
review: RORO 패턴 적용,  체크 표시 UI 테스트

RORO 패턴 적용, 체크 표시 UI 테스트 구현

sejineeee committed:
review: 각 리듀서 테스트 구현, 테스트명 변경

저장한다에서 변경한다로 변경
selected에 대한 테스트 구현
action이 없을 경우, 정의된 action type이 없을 경우 테스트 추가

sejineeee committed:
test: selector에 key 이름 변경

restaurants key 이름을 selectedRestaurant로 변경

sejineeee committed:
toBe에서 toHaveLength로 변경

toBe 사용하니 오류 발생하여 toHaveLength로 변경

sejineeee committed:
test: async action 테스트 추가

async action에 대한 테스트 시도
waitFor를 이용해보려고 했으나 실패

sejineeee committed:
test: api test 추가

fetch api 테스트 추가

sejineeee committed:
style: 포맷팅

sejineeee committed:
사용하지 않는 fixtures 데이터 제거

sejineeee commented:
## 내용
안녕하세요! 마지막인데...async action에 대한 테스트를 마무리 못 해서 너무 아쉽습니다!
일단 구현해보려고 해봤으나, 에러가 많이 났습니다... 😭
![스크린샷 2022-11-13 오후 5 38 25](https://user-images.githubusercontent.com/86041335/201513264-7c4628ab-5b03-478f-9c79-deb9ec826d97.png)
![스크린샷 2022-11-13 오후 5 38 18](https://user-images.githubusercontent.com/86041335/201513265-452bf5db-f48d-4ec5-b8c8-27733c00a058.png)

다른 분들의 코드를 보고 이해해보고 싶었으나... 이해가 잘 되지 않아서 문제예요!
혹시 비동기를 test 하기 위해서는 어떤 키워드를 알아야 되는지, 혹시 좋은 글이 있다면 공유해주시면 감사하겠습니다!!
이번 한 주도 정말 감사합니다!! 🙌🏻

daadaadaah commented on src/services/api.test.js:
```diff
@@ -0,0 +1,53 @@
+describe('api', () => {
+  describe('fetchRegions', () => {
+    beforeEach(() => {
+      global.fetch = jest.fn(() => Promise.resolve({
+        json: () => Promise.resolve(regions),
+      }));
```
중복되는 코드를 어떤 걸 활용해서 묶어줄 수 있을까여?ㅎㅎ

daadaadaah reviewed (changes requested):
- 이번주도 수고 많으셨어요! 비록 과제는 완성하지 못했지만 회고 쓰신 점 너무 잘하셨습니당!👍 다음주에도 같이 힘내봐요! 👊

> 다른 분들의 코드를 보고 이해해보고 싶었으나... 이해가 잘 되지 않아서 문제예요!
혹시 비동기를 test 하기 위해서는 어떤 키워드를 알아야 되는지, 혹시 좋은 글이 있다면 공유해주시면 감사하겠습니다!!
- 다른 분들 코드 중에서 redux-mock-store도 있어서, 이해하기 어려울 수 있는데, 예전에 제가 한 [코드]( https://github.com/CodeSoom/react-week5-assignment-1/pull/9/files#diff-0c891c925105fe84e46525d7c883d3abee5142b2177b964bf1060126086c9a47
) 한번  보실래여? redux-mock-store 안쓰고 했거든요!
